README.md
# stream-helper
Stream helper program for SSBM. Written for UMDsmash.

## Current Version
1.0

[changelog](CHANGES.md)

## Requirements
python 2.7

https://www.python.org/download/releases/2.7/

Qt 4.8.5 VS2010

https://download.qt.io/archive/qt/4.8/4.8.5/

PyQt 4

https://www.riverbankcomputing.com/software/pyqt/download

PILLOW

Run `python -m pip install pillow` in cmd

## Quick Start
1. Download or clone the repository.
2. Run `py_stream_helper.py`.
3. Set your streaming program to read the text and image files in the output directory.
4. Edit the values of the text boxes.
5. Click the save button.

The displayed text in the streaming program should be updated to the text in the helper program.

## Documentation
### Changing the text output files
Edit line 8 of `py_stream_helper.py` (the `text_folder` variable) to change the directory to the desired location.
The program will create the directory on startup if it does not exist.

Edit the filenames starting at line 115 to change the text file each text box will output to.

### Changing the destination folder of the currently selected images
Change the folder specified by `image_dest_folder` on line 10 to the folder you wish images to be copied to.

### Changing the filename of the currently selected images
Change the name contained in `image_dest_prefix` on line 11 to the filename you wish to use.
The numbers 0-3 and `image_suffix` will be appended to the filename.

### Changing the location of the image source folder
Change the folder specified by `sprite_folder` on line 9 to the folder containing the images you wish to select from.

### Changing the images used by the dropbox boxes
Replace the files in the folder identified on line 9 by `sprite_folder` with the images you wish to select from.
These images must be the format specified on line 12 by `image_suffix`.

### Changing the method used to create the light versions of images
The function `ConvertImage` beginning on line 22 of `py_stream_helper.py` converts an image to the lightened version.
By default it works by cutting the saturation of the image in half.
To change the fraction the saturation is divided by, change line 30: `s = s / 2.0` to another equation.
More advanced operations could be performed by replacing the entire function.

## Default values
text_folder = "output/"

sprite_folder = "SSBMsprites/"

image_dest_folder = "output/img/"

image_dest_prefix = "char_"

image_suffix = ".png"

CHANGES.md
# Version History

## 1.0
* Added title, player, and score text boxes
* Added swap button for players

## .2
* Added charcter image select boxes
* Added reset and save buttons
* Added autosave option

## .1
* Initial repo setup


character_heads.py
import colorsys
import os
import sys
import shutil
from PIL import Image
from PyQt4 import QtCore, QtGui

sprite_folder = "SSBMsprites/"
dest_prefix = "output/p"
file_suffix = ".png"

def ConvertImage(filename):
  im = Image.open(filename)
  ld = im.load()
  width, height = im.size
  for y in range(height):
    for x in range(width):
      r,g,b,a = ld[x,y]
      h,s,v = colorsys.rgb_to_hsv(r/255., g/255., b/255.)
      s = s/2.0
      r,g,b = colorsys.hsv_to_rgb(h, s, v)
      ld[x,y] = (int(r * 255.9999), int(g * 255.9999), int(b * 255.9999), a)
  im.save(filename)
      
class CharHeadComboBox(QtGui.QComboBox):
  def __init__(self, parent, num, checkbox):
    super(CharHeadComboBox, self).__init__(parent)
    self.num = num
    self.checkbox = checkbox
    self.autosave = False

  def Save(self):
    src = sprite_folder + str(self.currentText())
    dest = dest_prefix + str(self.num) + file_suffix
    #print "copied", src, "to", dest
    shutil.copyfile(src, dest)
    if self.checkbox.isChecked():
      #print "adjusting", dest
      ConvertImage(dest)
      #print "done"
    
  def onIndexChanged(self, index):
    if self.autosave:
      self.Save()

  def checkboxChanged(self, i):
    self.onIndexChanged(self.currentIndex())

class CharHeadWindow(QtGui.QWidget):
  def __init__(self):
    super(CharHeadWindow, self).__init__()
    self.initUI()

  def initUI(self):
    self.setGeometry(300, 300, 500, 120)
    self.setWindowTitle("Character Head Helper")

    # Grid layout
    grid = QtGui.QGridLayout()
    self.setLayout(grid)

    # Create comboboxes and checkboxes.
    self.combo_boxes = [None]*4
    self.check_boxes = [None]*4
    files = os.listdir(sprite_folder)
    empty_index = files.index("Empty" + file_suffix)
    for i in xrange(4):
      self.check_boxes[i] = QtGui.QCheckBox("light colors", self)
      grid.addWidget(self.check_boxes[i], 1, i)
    
      self.combo_boxes[i] = CharHeadComboBox(self, i, self.check_boxes[i])
      self.combo_boxes[i].addItems(files)
      self.combo_boxes[i].connect(self.combo_boxes[i], QtCore.SIGNAL("currentIndexChanged(int)"),
                                  self.combo_boxes[i].onIndexChanged)
      self.combo_boxes[i].setCurrentIndex(empty_index)
      grid.addWidget(self.combo_boxes[i], 0, i)

      self.check_boxes[i].connect(self.check_boxes[i], QtCore.SIGNAL("stateChanged(int)"),
                                  self.combo_boxes[i].checkboxChanged)
                             
    # Reset button
    self.reset_btn = QtGui.QPushButton("Reset", self)
    self.reset_btn.connect(self.reset_btn, QtCore.SIGNAL('clicked()'), self.Reset)
    grid.addWidget(self.reset_btn, 2, 0)

    # Autosave checkbox
    self.autosave_box = QtGui.QCheckBox("autosave", self)
    self.autosave_box.connect(self.autosave_box, QtCore.SIGNAL("stateChanged(int)"),
                              self.AutosaveChanged)
    grid.addWidget(self.autosave_box, 2, 1)
    
    # Save button
    save_btn = QtGui.QPushButton("Save", self)
    save_btn.connect(save_btn, QtCore.SIGNAL('clicked()'), self.Save)
    grid.addWidget(save_btn, 2, 2)
    
  def Reset(self):
    empty_index = self.combo_boxes[0].findText("Empty" + file_suffix, QtCore.Qt.MatchFixedString)
    for combo_box in self.combo_boxes:
      combo_box.setCurrentIndex(empty_index)
      
    for check_box in self.check_boxes:
      check_box.setCheckState(False)
      
  def AutosaveChanged(self, i):
    state = self.autosave_box.isChecked()
    for combo_box in self.combo_boxes:
      combo_box.autosave = state
      if state:
        combo_box.Save()
      
  def Save(self):
    for combo_box in self.combo_boxes:
      combo_box.Save()
      
def main():
  # Create an PyQT4 application object.
  a = QtGui.QApplication(sys.argv)
  w = CharHeadWindow()

  # Show window
  w.show()

  sys.exit(a.exec_())

if __name__ == "__main__":
  main()

py_stream_helper.py
import colorsys
import os
import sys
import shutil
from PIL import Image
from PyQt4 import QtCore, QtGui

text_folder = "output/"
sprite_folder = "SSBMsprites/"
image_dest_folder = "output/img/"
image_dest_prefix = "char_"
image_suffix = ".png"

def CreateDirectories():
  if not os.path.exists(text_folder):
    print "Creating text output folder", text_folder
    os.makedirs(text_folder)
  if not os.path.exists(image_dest_folder):
    print "Creating image output folder", image_dest_folder
    os.makedirs(image_dest_folder)

def ConvertImage(filename):
  im = Image.open(filename)
  ld = im.load()
  width, height = im.size
  for y in range(height):
    for x in range(width):
      r, g, b, a = ld[x,y]
      h, s, v = colorsys.rgb_to_hsv(r / 255.0, g / 255.0, b / 255.0)
      s = s / 2.0
      r, g, b = colorsys.hsv_to_rgb(h, s, v)
      ld[x,y] = (int(r * 255.9999), int(g * 255.9999), int(b * 255.9999), a)
  im.save(filename)

class AutoTextBox(QtGui.QLineEdit):
  def __init__(self, parent, other, filepath, placeholder):
    super(AutoTextBox, self).__init__(parent)
    self.other = other
    self.filepath = filepath
    self.setPlaceholderText(placeholder)
    self.autosave = False

    self.connect(self, QtCore.SIGNAL("textChanged(QString)"), self.OnChanged)

  def Save(self):
    with open(self.filepath, "w") as f:
      f.write(self.text())

  def SetAutosave(self, state):
    self.autosave = state
    if self.autosave:
      self.Save()

  def OnChanged(self, text):
    if self.autosave:
      self.Save()

class SwapButton(QtGui.QPushButton):
  def __init__(self, parent, text, pairs):
    super(SwapButton, self).__init__(text, parent)
    self.pairs = pairs

    self.connect(self, QtCore.SIGNAL("clicked()"), self.Swap)

  def Swap(self):
    for pair in self.pairs:
      temp_text = pair[0].text()
      pair[0].setText(pair[1].text())
      pair[1].setText(temp_text)

class CharHeadComboBox(QtGui.QComboBox):
  def __init__(self, parent, dest_path, num, files, check_box):
    super(CharHeadComboBox, self).__init__(parent)
    self.dest_path = dest_path
    self.num = num
    self.check_box = check_box
    self.autosave = False

    self.addItems(files)
    self.empty_index = self.findText("Empty" + image_suffix, QtCore.Qt.MatchFixedString)
    self.setCurrentIndex(self.empty_index)
    self.connect(self, QtCore.SIGNAL("currentIndexChanged(int)"), self.OnIndexChanged)
    check_box.connect(check_box, QtCore.SIGNAL("stateChanged(int)"), self.CheckboxChanged)

  def Save(self):
    src = sprite_folder + str(self.currentText())
    shutil.copyfile(src, self.dest_path)
    if self.check_box.isChecked():
      ConvertImage(self.dest_path)

  def SetAutosave(self, state):
    self.autosave = state
    if self.autosave:
      self.Save()

  def OnIndexChanged(self, index):
    if self.autosave:
      self.Save()

  def CheckboxChanged(self, i):
    self.OnIndexChanged(self.currentIndex())

class CharHeadWindow(QtGui.QWidget):
  def __init__(self):
    super(CharHeadWindow, self).__init__()
    self.initUI()

  def initUI(self):
    self.setGeometry(300, 300, 500, 240)
    self.setWindowTitle("PyStreamHelper")
    self.saveable_items = []

    # Title textboxes.
    self.text_boxes = [None]*6
    self.text_boxes[0] = AutoTextBox(self, self.text_boxes[1], text_folder + "title1.txt", "Title 1")
    self.text_boxes[1] = AutoTextBox(self, self.text_boxes[0], text_folder + "title2.txt", "Title 2")

    # Player name and score textboxes.
    self.text_boxes[4] = AutoTextBox(self, self.text_boxes[1], text_folder + "score_l.txt", "")
    self.text_boxes[4].setFixedWidth(30)
    self.text_boxes[2] = AutoTextBox(self, self.text_boxes[1], text_folder + "player_l.txt", "Player L")
    self.text_boxes[3] = AutoTextBox(self, self.text_boxes[1], text_folder + "player_r.txt", "Player R")
    self.text_boxes[5] = AutoTextBox(self, self.text_boxes[1], text_folder + "score_r.txt", "")
    self.text_boxes[5].setFixedWidth(30)

    self.saveable_items.extend(self.text_boxes)

    # Comboboxes and checkboxes.
    self.combo_boxes = [None]*4
    self.check_boxes = [None]*4
    files = os.listdir(sprite_folder)
    for i in xrange(4):
      self.check_boxes[i] = QtGui.QCheckBox("light colors", self)
      self.combo_boxes[i] = CharHeadComboBox(self, image_dest_folder + image_dest_prefix + str(i) + image_suffix,
                                             i, files, self.check_boxes[i])
    self.saveable_items.extend(self.combo_boxes)

    # Reset button.
    self.reset_btn = QtGui.QPushButton("Reset", self)
    self.reset_btn.connect(self.reset_btn, QtCore.SIGNAL('clicked()'), self.Reset)

    # Autosave checkbox.
    self.autosave_box = QtGui.QCheckBox("autosave", self)
    self.autosave_box.connect(self.autosave_box, QtCore.SIGNAL("stateChanged(int)"),
                              self.AutosaveChanged)

    # Save button.
    save_btn = QtGui.QPushButton("Save", self)
    save_btn.connect(save_btn, QtCore.SIGNAL('clicked()'), self.Save)

    # Swap button.
    self.swap_button = SwapButton(self, "Swap Players", ((self.text_boxes[2], self.text_boxes[3]),
                                                         (self.text_boxes[4], self.text_boxes[5])))

    # Add everything to the layout.
    main = QtGui.QVBoxLayout()
    
    # Row 0.
    row = QtGui.QHBoxLayout()
    row.addWidget(self.text_boxes[0])
    row.addWidget(self.text_boxes[1])
    main.addLayout(row)
    
    # Row 1.
    row = QtGui.QHBoxLayout()
    row.addWidget(self.text_boxes[4])
    row.addWidget(self.text_boxes[2])
    row.addWidget(self.text_boxes[3])
    row.addWidget(self.text_boxes[5])
    main.addLayout(row)
    
    # Rows 2+3.
    row1 = QtGui.QHBoxLayout()
    row2 = QtGui.QHBoxLayout()
    for i in xrange(4):
      row1.addWidget(self.combo_boxes[i])
      row2.addWidget(self.check_boxes[i])
    main.addLayout(row1)
    main.addLayout(row2)
      
    # Row 4.
    row = QtGui.QHBoxLayout()
    row.addWidget(self.reset_btn)
    row.addWidget(self.autosave_box)
    row.addWidget(save_btn)
    row.addWidget(self.swap_button)
    main.addLayout(row)
    self.setLayout(main)

  def Reset(self):
    for text_box in self.text_boxes:
      text_box.setText("")
    for combo_box in self.combo_boxes:
      combo_box.setCurrentIndex(combo_box.empty_index)
    for check_box in self.check_boxes:
      check_box.setCheckState(False)

  def AutosaveChanged(self, i):
    state = self.autosave_box.isChecked()
    for item in self.saveable_items:
      item.SetAutosave(state)

  def Save(self):
    for item in self.saveable_items:
      item.Save()

def main():
  CreateDirectories()

  # Create an PyQT4 application object.
  a = QtGui.QApplication(sys.argv)
  w = CharHeadWindow()

  # Show window
  w.show()

  sys.exit(a.exec_())

if __name__ == "__main__":
  main()